environment/core/util/distributions.hpp

```diff
@@ -0,0 +1,474 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE
+// SOFTWARE.
+
+// ==============================================================================
+
+// Copyright (c) 2009-2014 by the contributors listed in CREDITS.TXT
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include <cstdint>
+#include <limits>
+#include <type_traits>
+
+namespace util {
+    // Precondition:  __x != 0
+    static unsigned __clz(unsigned __x) {
+#ifndef _LIBCPP_COMPILER_MSVC
+        return static_cast<unsigned>(__builtin_clz(__x));
+#else
+        static_assert(sizeof(unsigned) == sizeof(unsigned long), "");
+        static_assert(sizeof(unsigned long) == 4, "");
+        unsigned long where;
+        // Search from LSB to MSB for first set bit.
+        // Returns zero if no set bit is found.
+        if (_BitScanReverse(&where, mask))
+            return 31 - where;
+        return 32; // Undefined Behavior.
+#endif
+    }
+
+    static unsigned long __clz(unsigned long __x) {
+#ifndef _LIBCPP_COMPILER_MSVC
+        return static_cast<unsigned long>(__builtin_clzl (__x));
+#else
+        static_assert(sizeof(unsigned) == sizeof(unsigned long), "");
+        return __clz(static_cast<unsigned>(__x));
+#endif
+    }
+
+    static unsigned long long __clz(unsigned long long __x) {
+#ifndef _LIBCPP_COMPILER_MSVC
+        return static_cast<unsigned long long>(__builtin_clzll(__x));
+#else
+        unsigned long where;
+        // BitScanReverse scans from MSB to LSB for first set bit.
+        // Returns 0 if no set bit is found.
+#if defined(_LIBCPP_HAS_BITSCAN64)
```

Comment:
Ok to leave undefined. We do support building for 32 bit windows, which lacks _BitScanReverse64. Don't care about performance implications here.